airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py

```diff
@@ -8,62 +8,145 @@
-    dockerd_container = (
+    apk_packages_to_install = [
+        STORAGE_DRIVER,
+        # Curl is only used for debugging purposes.
+        "curl",
+    ]
+    base_container = (
         dagger_client.container()
         .from_(consts.DOCKER_DIND_IMAGE)
         # We set this env var because we need to use a non-default zombie reaper setting.
         # The reason for this is that by default it will want to set its parent process ID to 1 when reaping.
         # This won't be possible because of container-ception: dind is running inside the dagger engine.
         # See https://github.com/krallin/tini#subreaping for details.
         .with_env_variable("TINI_SUBREAPER", "")
-        # Similarly, because of container-ception, we have to use the fuse-overlayfs storage engine.
         .with_exec(
             sh_dash_c(
                 [
-                    # Update package metadata.
                     "apk update",
-                    # Install the storage driver package.
-                    "apk add fuse-overlayfs",
-                    # Update daemon config with storage driver.
+                    f"apk add {' '.join(apk_packages_to_install)}",
                     "mkdir /etc/docker",
-                    '(echo {\\"storage-driver\\": \\"fuse-overlayfs\\"} > /etc/docker/daemon.json)',
                 ]
             )
         )
         # Expose the docker host port.
         .with_exposed_port(DOCKER_HOST_PORT)
-        # Mount the docker cache volumes.
-        .with_mounted_cache("/var/lib/docker", dagger_client.cache_volume(DOCKER_VAR_LIB_VOLUME_NAME))
+        # We cache /tmp for file sharing between client and daemon.
         .with_mounted_cache("/tmp", dagger_client.cache_volume(DOCKER_TMP_VOLUME_NAME))
     )
-    if docker_hub_username_secret and docker_hub_password_secret:
-        dockerd_container = (
+    if docker_var_lib_caching:
+        # We cache /var/lib/docker to avoid downloading images and layers multiple times.
+        base_container = base_container.with_mounted_cache("/var/lib/docker", dagger_client.cache_volume(DOCKER_VAR_LIB_VOLUME_NAME))
+    return base_container
+
+
+def get_daemon_config_json(registry_mirror_url: Optional[str] = None) -> str:
+    """Get the json representation of the docker daemon config.
+
+    Args:
+        registry_mirror_url (Optional[str]): The registry mirror url to use.
+
+    Returns:
+        str: The json representation of the docker daemon config.
+    """
+    daemon_config = {
+        "storage-driver": STORAGE_DRIVER,
+    }
+    if registry_mirror_url:
+        daemon_config["registry-mirrors"] = ["http://" + registry_mirror_url]
+        daemon_config["insecure-registries"] = [registry_mirror_url]
+    return json.dumps(daemon_config)
+
+
+def docker_login(
+    dockerd_container: Container,
+    docker_registry_username_secret: Optional[Secret],
+    docker_registry_password_secret: Optional[Secret],
+    docker_registry_address: Optional[str] = "docker.io",
```

Comment:
Similar comment regarding function arg that doesn't really serve a purpose. Here a constant would be better.

Reply:
moved to a const